test/LockedNORI.t.sol

```diff
@@ -0,0 +1,191 @@
+pragma solidity 0.8.13;
+
+import "forge-std/Test.sol";
+import "forge-std/Vm.sol";
+import "../contracts/BridgedPolygonNORI.sol";
+import "../contracts/LockedNORI.sol";
+import "@openzeppelin/contracts-upgradeable/interfaces/IERC1820RegistryUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC777/IERC777RecipientUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC777/IERC777SenderUpgradeable.sol";
+
+abstract contract ERC777ERC1820 is
+  IERC777RecipientUpgradeable,
+  IERC777SenderUpgradeable
+{
+  function tokensReceived(
+    address operator,
+    address from,
+    address to,
+    uint256 amount,
+    bytes calldata userData,
+    bytes calldata operatorData
+  ) external {
+    console.log("tokensReceived");
+  }
+
+  function tokensToSend(
+    address operator,
+    address from,
+    address to,
+    uint256 amount,
+    bytes calldata userData,
+    bytes calldata operatorData
+  ) external {
+    console.log("tokensToSend");
+  }
+}
+
+contract Recipient {
+  address constant _ERC1820_REGISTRY_ADDRESS =
+    0x1820a4B7618BdE71Dce8cdc73aAB6C95905faD24;
+  uint256 counter;
+  bool shouldAttack;
+  LockedNORI lNori;
+  BridgedPolygonNORI bpNori;
+
+  constructor(bool attack) {
+    counter = 0;
+    shouldAttack = attack;
+    IERC1820RegistryUpgradeable registry = IERC1820RegistryUpgradeable(
+      _ERC1820_REGISTRY_ADDRESS
+    );
+    registry.setInterfaceImplementer(
+      address(this),
+      keccak256("ERC777TokensSender"),
+      address(this)
+    );
+    registry.setInterfaceImplementer(
+      address(this),
+      keccak256("ERC777TokensRecipient"),
+      address(this)
+    );
+  }
+
+  function tokensReceived(
+    address operator,
+    address from,
+    address to,
+    uint256 amount,
+    bytes calldata userData,
+    bytes calldata operatorData
+  ) external {
+    counter = counter + 1;
+    if (counter == 1) {
+        bpNori = BridgedPolygonNORI(operator);
+    }
+    console.log("tokensReceived - recipient contract, call count", counter);
+    console.log('operator %s, amount %d', operator, amount);
+    if (counter == 2 && shouldAttack) {
+      lNori = LockedNORI(operator);
+      console.log("Re-entry attempt");
+      lNori.withdrawTo(address(this), amount);
+    }
```

Comment:
@scottynomad 
iirc it's the tokensToSend function that is the exploitable one-- so not sure if this is testing this in the right place. Might be remembering wrong, but that's what I did here : [victim](https://github.com/jaycenhorton/reenter-777/blob/main/contracts/Victim.sol#L42), [attacker](https://github.com/jaycenhorton/reenter-777/blob/main/contracts/Attacker.sol#L43)

Reply:
:100: for getting foundry semi-configured here though!

Reply:
Yeah, I think both are potentially exploitable depending on where the state update logic is placed.  Fortunately the implementation of `LockedNORI.withdrawTo` resembles `__buyShareFromUser_workaround` more than the vulnerable version.

Reply:
😅